angrydd.6
.\" Copyright 2004 Joe Wreschnig <piman@sacredchao.net>
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" version 2 as published by the Free Software Foundation.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139,
.\" USA.
.\"
.\" $Id: angrydd.6 317 2006-01-12 21:19:27Z piman $
.TH angrydd 6 "July 29th, 2004" SCOL "Angry, Drunken Dwarves Manual"
.SH NAME
angrydd \- a falling block puzzle game
.SH SYNOPSIS
angrydd [ \fB\-\-debug\fR ]
.SH DESCRIPTION
In Angry, Drunken Dwarves, you are an angry, drunken dwarf. Why are you so
angry? Who knows. But you've decided to take your aggression out on other
dwarves, by dropping gems on their heads. Lots of gems.
.PP
You have your own supply of gems, which will slowly fall from
the top of the screen. As gems reach the bottom, they will land and sit in
place. If you put enough gems of the same color in a rectangle shape, they
will merge and form a crystal.
.PP
Less common are break gems, which are circular. When you drop a break gem
on a gem or crystal of the same color, it will destroy any gems of that
color adjacent to it. Crystals are worth much more than an equal number of
regular gems.
.PP
For more information, see the "How to Play" screen in the game.
.SH OPTIONS
.IP \-\-debug
Start the game without Psyco support, for debugging and profiling
purposes.
.IP \-\-help,\ \-h
Display a brief help message.
.IP \-\-version,\ \-v
Display copyright and version information.
.SH USAGE
Player 1's controls are w, a, s, d for up, left, down, right, q for rotate
counter-clockwise, e for clockwise, and Enter for start. Player 1 can also
use the arrow keys to move.
.PP
Player 2's controls are i, j, k, l for up, left, down, right, u for rotate
counter-clockwise, o for clockwise, and backspace for start.
.PP
Pressing 'p' during a game will pause it. Pressing escape will usually
take you back to the menu.
.PP
The game is designed to be played with joysticks or gamepads as well;
the directionals will move, the buttons rotate, start is start,
and select is pause.
.SH GAME MODES
.IP Dwarf\ versus\ Dwarf
Two players fight against each other. You can select the number of matches
and other options in the Setup menu.
.IP Arcade\ Mode
Go up against each dwarf in game, each one harder than the previous.
.IP Dwarf\ versus\ Robot
To get this mode, you must do well in Arcade Mode. In this mode you can go
up against any dwarf you want played by an AI, and also set the AI
level in the Setup menu. Select the AI's character with j and l, and
confirm it with 2.
.IP Others?
There are other, secret game modes you can unlock. The "How to Play"
screen will give you instructions on these modes when you can play them.
.SH AUTHOR
Joe Wreschnig <piman@sacredchao.net> is the author of Angry, Drunken Dwarves.


debian/changelog
angrydd (1.0.2-1) hardy; urgency=low

  *whatever 

 -- Josh Cronemeyer <joshuacronemeyer@shellsink.com>  Sun, 19 Jul 2009 11:06:10 -0700

angrydd (1.0.1-9.1) hardy; urgency=low

  *rebuilding since launchpad ignored my last build 

 -- Josh Cronemeyer <joshuacronemeyer@shellsink.com>  Sun, 19 Jul 2009 10:07:55 -0700

angrydd (1.0.1-9) hardy; urgency=low

  *add pairing monster 

 -- Josh Cronemeyer <joshuacronemeyer@shellsink.com>  Sat, 18 Jul 2009 20:57:51 -0700

angrydd (1.0.1-6) unstable; urgency=low

  * debian/menu: Updated longtitle and section.  Closes: #445123.
  * debian/angrydd.desktop: Updated categories.

 -- Bart Martens <bartm@knars.be>  Wed, 03 Oct 2007 21:13:03 +0200

angrydd (1.0.1-5) unstable; urgency=low

  * debian/patches/03_bug405368.diff: Documented keys.  Closes: #405368.
  * debian/patches/04_bug406548.diff: Documented the fullscreen toggle in the
    manpage, and added this toggle to the config menu.  Closes: #406548.
  * debian/copyright: Updated.

 -- Bart Martens <bartm@knars.be>  Sat, 13 Jan 2007 18:37:16 +0100

angrydd (1.0.1-4) unstable; urgency=low

  * debian/patches/02_unixbros.diff: Breaking a crystal of "over 16 gems"
    unlocks "unixbros".  Closes: #403023.
  * debian/angrydd.xpm, debian/menu: Added menu icon and long title.

 -- Bart Martens <bartm@knars.be>  Fri, 15 Dec 2006 22:00:13 +0100

angrydd (1.0.1-3) unstable; urgency=low

  * debian/control: Architecture "all".  Closes: #400537.

 -- Bart Martens <bartm@knars.be>  Mon, 27 Nov 2006 07:50:51 +0100

angrydd (1.0.1-2) unstable; urgency=low

  * debian/copyright: Fixed.  Closes: #400237.

 -- Bart Martens <bartm@knars.be>  Fri, 24 Nov 2006 18:44:56 +0100

angrydd (1.0.1-1) unstable; urgency=low

  * Initial release.  Closes: #334173.
  * debian/pyversions: Needs python 2.3 or newer.
  * debian/control: Needs python-pygame (>= 1.6.2).
  * debian/patches/01_prefix_usr.diff: Patches PREFIX in Makefile.
  * debian/angrydd.desktop, debian/install: Added.
  * debian/watch: Added.

 -- Bart Martens <bartm@knars.be>  Mon, 13 Nov 2006 15:39:44 +0100



angrydd.py
#!/usr/bin/env python

# angrydd - a falling block puzzle game
# Copyright 2004 Joe Wreschnig <piman@sacredchao.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# version 2 along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

__date__ = "$LastChangedDate: 2006-01-12 15:19:27 -0600 (Thu, 12 Jan 2006) $"
__version__   = "1.0.1"
__revision__  = "$Id: angrydd.py 317 2006-01-12 21:19:27Z piman $"
__copyright__ = "Copyright 2004 Joe Wreschnig <piman@sacredchao.net>"
__license__   = "GNU GPL v2"
__author__    = "Joe Wreschnig <piman@sacredchao.net>"

import os
import sys

def print_help():
    print """\
Usage: %s [ --debug | --help | --version ]
Options:
  --help, -h     Print this help message and exit.
  --version, -v  Print version and copyright information and exit.
  --debug        Start without Psyco support, for profiling and debugging.
  --profile      Start with profiling (implies --debug).
""" %(
        sys.argv[0])
    raise SystemExit

def print_version():
    print """\
Angry Drunken Programmers %s - a falling block puzzle game
Copyright 2004 %s

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU GPL version 2, as published
by the Free Software Foundation.""" % (__version__, __author__)
    raise SystemExit

if __name__ == "__main__":
    if ("--help" in sys.argv) or ("-h" in sys.argv): print_help()
    elif ("--version" in sys.argv) or ("-v" in sys.argv): print_version()

angrydd_path = os.path.split(os.path.realpath(__file__))[0]
if os.path.exists(os.path.join(angrydd_path, "angrydd.zip")):
    sys.path.insert(0, os.path.join(angrydd_path, "angrydd.zip"))

import pygame
from pygame import mixer

from constants import *

import config
import load

# Toggling fullscreen doesn't work on non-X11 platforms as of Pygame 1.7.
def toggle_fs():
    flags = 0
    fs = config.getboolean("settings", "fullscreen") ^ True
    if "linux" not in sys.platform:
        if fs: flags ^= FULLSCREEN
        pygame.display.set_mode([800, 600], flags)
    else: pygame.display._toggle_fullscreen()
    config.set("settings", "fullscreen", str(fs))

pygame.display._toggle_fullscreen = pygame.display.toggle_fullscreen
pygame.display.toggle_fullscreen = toggle_fs

def main():
    from characters import Character

    init()

    import menu
    import howtoplay
    import unlocker
    main_menu = [
        ("Dwarf versus Dwarf", menu.entry(play_game, False)),
        ("Arcade Mode", menu.entry(play_arcade_game, False)),
        ("Setup", menu.entry(setup, False)),
        ("How to Play", menu.entry(howtoplay.init, False)),
        ("Quit Game", menu.break_menu)]

    if config.getboolean("unlock", "unlocker"):
        main_menu.insert(3, ("View Unlocks", menu.entry(unlocker.init, False)))

    if config.getboolean("unlock", "single"):
        main_menu.insert(2, ("Single Player",
                             menu.entry(play_single_game, False)))
    if config.getboolean("unlock", "cpuversus"):
        main_menu.insert(2,
                         ("Dwarf versus Robot",
                          menu.entry(play_ai_game, False)))

#    if len(Character.arcade) != 8:
#        main_menu = [main_menu[0]] + main_menu[2:]

    menu.Menu(main_menu)
    quit()

# Start a player vs. player game
def play_game(menu, platform, pos, key):
    import game, charselect
    characters = charselect.init()
    if characters[0] is not None:
        game.VersusGame(characters)
        mixer.music.load(os.path.join("music", "intro.ogg"))
        mixer.music.play(-1)
    return True

# Start a single player game
def play_single_game(menu, platform, pos, key):
    import game
    game.SingleGame()
    mixer.music.load(os.path.join("music", "intro.ogg"))
    mixer.music.play(-1)
    return True

# Start a game versus an AI
def play_ai_game(menu, platform, pos, key):
    import game, charselect
    characters = charselect.init()
    if characters[0] is not None:
        game.AIGame(characters)
        mixer.music.load(os.path.join("music", "intro.ogg"))
        mixer.music.play(-1)
    return True

def play_arcade_game(menu, platform, pos, key):
    import game, charselect
    characters = charselect.init(numplayers = 1)
    if characters[0] is not None:
        game.ArcadeGame(characters)
        mixer.music.load(os.path.join("music", "intro.ogg"))
        mixer.music.play(-1)
    return True

# Open the configuration menu
def setup(menu_, platform, pos, key):
    import menu
    config_menu = [
        (config.get_matches(), menu.entry(config.set_matches)),
        (config.get_speed(), menu.entry(config.set_speed)),
        (config.get_rotup(), menu.entry(config.set_rotup)),
        (config.get_space(), menu.entry(config.set_space)),

        ("Back", menu.break_menu)
        ]

    if config.getboolean("unlock", "cpuversus"):
        config_menu.insert(2, (config.get_ai(), menu.entry(config.set_ai)))

    if config.getboolean("unlock", "combat"):
        config_menu.insert(-1, (config.get_combat(),
                                menu.entry(config.set_combat)))

    try: menu.Menu(config_menu)
    except menu.MenuExit: pass
    return True

# Set up Pygame, load sounds, etc.
def init():
    if os.name == 'posix': # We need to force stereo in many cases.
        try: mixer.pre_init(44100, -16, 2)
        except pygame.error: pass

    pygame.init()
    config.init(rc_file)
    os.chdir(angrydd_path)
    pygame.display.set_caption("Angry, Drunken Programmers")
    try: pygame.display.set_icon(pygame.image.load("angrydd.png"))
    except pygame.error: pass
    pygame.mouse.set_visible(False)
    if config.getboolean("settings", "fullscreen"): flags = FULLSCREEN
    else: flags = 0
    pygame.display.set_mode([800, 600], flags)

    import game
    import menu
    import boxes
    import characters

    boxes.TickBox.sound = load.sound("tick.wav")
    boxes.TickBox.sound.set_volume(0.3)

    boxes.BreakBox.sound = load.sound("break.wav")
    boxes.BreakBox.sound.set_volume(0.3)

    boxes.Special.sound = load.sound("select-confirm.wav")
    boxes.Special.sound.set_volume(0.5)

    game.FallingBoxes.rotate_sound = load.sound("rotate.wav")
    game.FallingBoxes.rotate_sound.set_volume(0.2)

    menu.Menu.bg = load.image("menu-bg.png").convert()

    characters.init()

    mixer.music.load(os.path.join("music", "intro.ogg"))
    mixer.music.play(-1)

# Shut down Pygame, save configuration data, turn on the Unlocker.
def quit(*args):
    pygame.quit()
    print "Saving configuration."
    config.unlock("unlocker")
    if int(config.get("scores-versus-3", "1").split(",")[1]) > 20000:
        config.unlock("single")
    config.write(file(rc_file, "w"))
    raise SystemExit

if __name__ == "__main__":
    profile = ("--profile" in sys.argv)
    debug = profile or ("--debug" in sys.argv)
    double_ai = ("--double-ai" in sys.argv)

    if not debug:
        try:
            import psyco
            psyco.full()
            print "Psyco optimizing compiler found. Using psyco.full()."
        except ImportError: pass

    if profile:
        import profile
        profile.run("main()")
    elif double_ai:
        # Hack to run two AIs versus each other; use --double-ai AI1,Ai2
        # on the command line. Bypasses main(), very buggy, for testing
        # use only.
        init()
        ai1, ai2 = sys.argv[sys.argv.index("--double-ai") + 1].split(",")
        game.AIGame([Character.available[0], Character.available[0]],
                     all_ais = True, ai1 = ai1, ai2 = ai2)
    else: main()
